### PR TITLE
feat(exmo): createOrder - trigger orders and margin orders

### PR DESCRIPTION
**Pull request by samgermain**
```
 % exmo createOrder ADA/USDT market buy 10 --verbose
2023-09-07T00:27:08.495Z
Node.js: v18.15.0
CCXT v4.0.83
exmo.createOrder (ADA/USDT, market, buy, 10)
fetch Request:
 exmo POST https://api.exmo.com/v1.1/order_create 
RequestBody:
 nonce=1694046431941&pair=ADA_USDT&quantity=10&price=0&type=market_buy 

handleRestResponse:
 exmo POST https://api.exmo.com/v1.1/order_create 200 OK 
ResponseBody:
 {"result":true,"error":"","order_id":36682140978} 

2023-09-07T00:27:12.126Z iteration 0 passed in 537 ms

{
  id: '36682140978',
  clientOrderId: undefined,
  datetime: undefined,
  timestamp: undefined,
  lastTradeTimestamp: undefined,
  status: undefined,
  symbol: 'ADA/USDT',
  type: undefined,
  timeInForce: undefined,
  postOnly: undefined,
  side: undefined,
  price: undefined,
  stopPrice: undefined,
  triggerPrice: undefined,
  cost: undefined,
  amount: undefined,
  filled: undefined,
  remaining: undefined,
  average: undefined,
  trades: [],
  fee: undefined,
  info: { result: true, error: '', order_id: '36682140978' },
  fees: [],
  lastUpdateTimestamp: undefined,
  reduceOnly: undefined,
  takeProfitPrice: undefined,
  stopLossPrice: undefined
}
2023-09-07T00:27:12.126Z iteration 1 passed in 537 ms
```

```
% exmo createOrder ADA/USDT limit buy 10 0.24 --verbose
2023-09-07T00:27:58.918Z
Node.js: v18.15.0
CCXT v4.0.83
exmo.createOrder (ADA/USDT, limit, buy, 10, 0.24)
fetch Request:
 exmo POST https://api.exmo.com/v1.1/order_create 
RequestBody:
 nonce=1694046481719&pair=ADA_USDT&quantity=10&price=0.24&type=buy 

handleRestResponse:
 exmo POST https://api.exmo.com/v1.1/order_create 200 OK 
ResponseBody:
 {"result":true,"error":"","order_id":36682146478} 

2023-09-07T00:28:01.877Z iteration 0 passed in 510 ms

{
  id: '36682146478',
  clientOrderId: undefined,
  datetime: undefined,
  timestamp: undefined,
  lastTradeTimestamp: undefined,
  status: undefined,
  symbol: 'ADA/USDT',
  type: undefined,
  timeInForce: undefined,
  postOnly: undefined,
  side: undefined,
  price: undefined,
  stopPrice: undefined,
  triggerPrice: undefined,
  cost: undefined,
  amount: undefined,
  filled: undefined,
  remaining: undefined,
  average: undefined,
  trades: [],
  fee: undefined,
  info: { result: true, error: '', order_id: '36682146478' },
  fees: [],
  lastUpdateTimestamp: undefined,
  reduceOnly: undefined,
  takeProfitPrice: undefined,
  stopLossPrice: undefined
}
2023-09-07T00:28:01.877Z iteration 1 passed in 510 ms
```

```
% exmo createOrder ADA/USDT market sell 10 0.24 '{"stopPrice": 0.24}' --verbose
2023-09-07T00:33:07.162Z
Node.js: v18.15.0
CCXT v4.0.83
exmo.createOrder (ADA/USDT, market, sell, 10, 0.24, [object Object])
fetch Request:
 exmo POST https://api.exmo.com/v1.1/stop_market_order_create 
RequestBody:
 nonce=1694046790761&pair=ADA_USDT&quantity=10&type=sell&trigger_price=0.24 

handleRestResponse:
 exmo POST https://api.exmo.com/v1.1/stop_market_order_create 200 OK 
ResponseBody:
 {"client_id":0,"parent_order_id":507088114129720346,"parent_order_id_str":"507088114129720346"} 

2023-09-07T00:33:11.061Z iteration 0 passed in 652 ms

{
  id: undefined,
  clientOrderId: '0',
  datetime: undefined,
  timestamp: undefined,
  lastTradeTimestamp: undefined,
  status: undefined,
  symbol: 'ADA/USDT',
  type: undefined,
  timeInForce: undefined,
  postOnly: undefined,
  side: undefined,
  price: undefined,
  stopPrice: undefined,
  triggerPrice: undefined,
  cost: undefined,
  amount: undefined,
  filled: undefined,
  remaining: undefined,
  average: undefined,
  trades: [],
  fee: undefined,
  info: {
    client_id: '0',
    parent_order_id: '507088114129720346',
    parent_order_id_str: '507088114129720346'
  },
  fees: [],
  lastUpdateTimestamp: undefined,
  reduceOnly: undefined,
  takeProfitPrice: undefined,
  stopLossPrice: undefined
}
2023-09-07T00:33:11.061Z iteration 1 passed in 652 ms
```

```
 % exmo createOrder ADA/USDT limit buy 10 0.24 '{"marginMode": "cross", "leverage": 3, "postOnly": true}' --verbose
2023-09-07T00:43:50.596Z
Node.js: v18.15.0
CCXT v4.0.83
exmo.createOrder (ADA/USDT, limit, buy, 10, 0.24, [object Object])
fetch Request:
 exmo POST https://api.exmo.com/v1.1/margin/user/order/create 
RequestBody:
 nonce=1694047433234&pair=ADA_USDT&quantity=10&price=0.24&type=limit_buy&leverage=3 

handleRestResponse:
 exmo POST https://api.exmo.com/v1.1/margin/user/order/create 200 OK 
ResponseBody:
 {"order_id":"692862104443994005"} 

2023-09-07T00:43:53.402Z iteration 0 passed in 520 ms

{
  id: '692862104443994005',
  clientOrderId: undefined,
  datetime: undefined,
  timestamp: undefined,
  lastTradeTimestamp: undefined,
  status: undefined,
  symbol: 'ADA/USDT',
  type: undefined,
  timeInForce: undefined,
  postOnly: undefined,
  side: undefined,
  price: undefined,
  stopPrice: undefined,
  triggerPrice: undefined,
  cost: undefined,
  amount: undefined,
  filled: undefined,
  remaining: undefined,
  average: undefined,
  trades: [],
  fee: undefined,
  info: { order_id: '692862104443994005' },
  fees: [],
  lastUpdateTimestamp: undefined,
  reduceOnly: undefined,
  takeProfitPrice: undefined,
  stopLossPrice: undefined
}
2023-09-07T00:43:53.402Z iteration 1 passed in 520 ms
```

```
 % exmo createOrder ADA/USDT market buy 10 undefined '{"marginMode": "cross", "leverage": 3, "stopPrice": 0.3}' --verbose
2023-09-07T00:52:02.270Z
Node.js: v18.15.0
CCXT v4.0.83
exmo.createOrder (ADA/USDT, market, buy, 10, , [object Object])
fetch Request:
 exmo POST https://api.exmo.com/v1.1/margin/user/order/create 
RequestBody:
 nonce=1694047924970&pair=ADA_USDT&quantity=10&stop_price=0.3&type=stop_buy&leverage=3 

handleRestResponse:
 exmo POST https://api.exmo.com/v1.1/margin/user/order/create 200 OK 
ResponseBody:
 {"order_id":"692862104444863062"} 

2023-09-07T00:52:05.134Z iteration 0 passed in 515 ms

{
  id: '692862104444863062',
  clientOrderId: undefined,
  datetime: undefined,
  timestamp: undefined,
  lastTradeTimestamp: undefined,
  status: undefined,
  symbol: 'ADA/USDT',
  type: undefined,
  timeInForce: undefined,
  postOnly: undefined,
  side: undefined,
  price: undefined,
  stopPrice: undefined,
  triggerPrice: undefined,
  cost: undefined,
  amount: undefined,
  filled: undefined,
  remaining: undefined,
  average: undefined,
  trades: [],
  fee: undefined,
  info: { order_id: '692862104444863062' },
  fees: [],
  lastUpdateTimestamp: undefined,
  reduceOnly: undefined,
  takeProfitPrice: undefined,
  stopLossPrice: undefined
}
2023-09-07T00:52:05.134Z iteration 1 passed in 515 ms
```

### CODE REVIEW

File: ts/src/exmo.ts [hunk 1]
```diff
@@ -1329,34 +1329,43 @@ export default class exmo extends Exchange {
...
          * @returns {object} an [order structure]{@link https://github.com/ccxt/ccxt/wiki/Manual#order-structure}
          */
         await this.loadMarkets ();
         const market = this.market (symbol);
-        const prefix = (type === 'market') ? (type + '_') : '';
-        const orderType = prefix + side;
         const isMarket = (type === 'market') && (price === undefined);
+        let marginMode = undefined;
+        [ marginMode, params ] = this.handleMarginModeAndParams ('createOrder', params);
+        const isSpot = (marginMode !== 'cross') && (marginMode !== 'isolated');
```
Comment (carlosmiei) [inline ts/src/exmo.ts:1350]: isn't easier to do `isSpot = (marginMode === undefined)`

File: ts/src/exmo.ts [hunk 2]
```diff
@@ -1365,16 +1374,59 @@ export default class exmo extends Exchange {
...
         }
-        if ((type === 'stop') || (type === 'stop_limit') || (type === 'trailing_stop')) {
-            const stopPrice = this.safeNumber2 (params, 'stop_price', 'stopPrice');
-            if (stopPrice === undefined) {
+        const leverage = this.safeNumber (params, 'leverage');
+        if (!isSpot && (leverage === undefined)) {
+            throw new ArgumentsRequired (this.id + ' createOrder requires an extra param params["leverage"] for margin orders');
+        }
+        params = this.omit (params, [ 'stopPrice', 'stop_price', 'triggerPrice', 'timeInForce', 'client_id', 'clientOrderId' ]);
+        if ((type === 'stop_buy') || (type === 'stop_sell') || (type === 'stop_limit_buy') || (type === 'stop_limit_sell') || (type === 'trailing_stop_buy') || (type === 'trailing_stop_sell')) {
```
Comment (carlosmiei) [inline ts/src/exmo.ts:1383]: this logic is a bit confusing, let's say the user provides `type = 'stop_limit_buy`, we just set the stop_price parameter? not even the type is put there?

File: ts/src/exmo.ts (hunk 1 above)
Comment (samgermain) [reply ts/src/exmo.ts:1350]: I wanted to leave the option of setting `marginMode` equal to `spot` also

Comment (carlosmiei) [reply ts/src/exmo.ts:1350]: @samgermain I think that is a bit weird and unnecessary

File: ts/src/exmo.ts (hunk 2 above)
Comment (samgermain) [reply ts/src/exmo.ts:1386]: I updated it

```
%  exmo createOrder ADA/USDT stop_sell sell 10 undefined '{"stopPrice": 0.24, "marginMode": "isolated", "leverage": 3}'
(node:3074) ExperimentalWarning: Custom ESM Loaders is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
2023-09-07T16:24:05.261Z
Node.js: v18.15.0
CCXT v4.0.83
exmo.createOrder (ADA/USDT, stop_sell, sell, 10, , [object Object])
2023-09-07T16:24:08.021Z iteration 0 passed in 510 ms

{
  id: '692862104547551001',
  clientOrderId: undefined,
  datetime: undefined,
  timestamp: undefined,
  lastTradeTimestamp: undefined,
  status: undefined,
  symbol: 'ADA/USDT',
  type: undefined,
  timeInForce: undefined,
  postOnly: undefined,
  side: undefined,
  price: undefined,
  stopPrice: undefined,
  triggerPrice: undefined,
  cost: undefined,
  amount: undefined,
  filled: undefined,
  remaining: undefined,
  average: undefined,
  trades: [],
  fee: undefined,
  info: { order_id: '692862104547551001' },
  fees: [],
  lastUpdateTimestamp: undefined,
  reduceOnly: undefined,
  takeProfitPrice: undefined,
  stopLossPrice: undefined
}
2023-09-07T16:24:08.021Z iteration 1 passed in 510 ms
```

File: ts/src/exmo.ts [hunk 3]
```diff
@@ -1329,34 +1329,42 @@ export default class exmo extends Exchange {
...
+         * @see https://documenter.getpostman.com/view/10287440/SzYXWKPi#3561b86c-9ff1-436e-8e68-ac926b7eb523  // margin
          * @param {string} symbol unified symbol of the market to create an order in
          * @param {string} type 'market' or 'limit'
          * @param {string} side 'buy' or 'sell'
          * @param {float} amount how much of currency you want to trade in units of base currency
          * @param {float} [price] the price at which the order is to be fullfilled, in units of the quote currency, ignored in market orders
          * @param {object} [params] extra parameters specific to the exmo api endpoint
+         * @param {float} [params.stopPrice] the price at which a trigger order is triggered at
+         * @param {string} [params.timeInForce] *spot only* 'fok', 'ioc' or 'post_only'
+         * @param {boolean} [params.postOnly] *spot only* true for post only orders
```
Comment (carlosmiei) [inline ts/src/exmo.ts:1342]: missing the error handling here

**Comment by carlosmiei** [comment]
LGTM
```
 p exmo createOrder "TON/USDT" "limit" buy 5 1   
Python v3.10.9
CCXT v4.0.83
exmo.createOrder(TON/USDT,limit,buy,5,1)
{'amount': None,
 'average': None,
 'clientOrderId': None,
 'cost': None,
 'datetime': None,
 'fee': None,
 'fees': [],
 'filled': None,
 'id': '36785559245',
 'info': {'error': '', 'order_id': '36785559245', 'result': True},
 'lastTradeTimestamp': None,
 'lastUpdateTimestamp': None,
 'postOnly': None,
 'price': None,
 'reduceOnly': None,
 'remaining': None,
 'side': None,
 'status': None,
 'stopLossPrice': None,
 'stopPrice': None,
 'symbol': 'TON/USDT',
 'takeProfitPrice': None,
 'timeInForce': None,
 'timestamp': None,
 'trades': [],
 'triggerPrice': None,
 'type': None}
```
```
 p exmo createOrder "TON/USDT" market buy 2   
Python v3.10.9
CCXT v4.0.83
exmo.createOrder(TON/USDT,market,buy,2)
{'amount': None,
 'average': None,
 'clientOrderId': None,
 'cost': None,
 'datetime': None,
 'fee': None,
 'fees': [],
 'filled': None,
 'id': '36785562966',
 'info': {'error': '', 'order_id': '36785562966', 'result': True},
 'lastTradeTimestamp': None,
 'lastUpdateTimestamp': None,
 'postOnly': None,
 'price': None,
 'reduceOnly': None,
 'remaining': None,
 'side': None,
 'status': None,
 'stopLossPrice': None,
 'stopPrice': None,
 'symbol': 'TON/USDT',
 'takeProfitPrice': None,
 'timeInForce': None,
 'timestamp': None,
 'trades': [],
 'triggerPrice': None,
 'type': None}
```
```
 p exmo createOrder "TON/USDT" market buy 2 None '{"triggerPrice": 4}'          
Python v3.10.9
CCXT v4.0.83
exmo.createOrder(TON/USDT,market,buy,2,None,{'triggerPrice': 4})
{'amount': None,
 'average': None,
 'clientOrderId': '0',
 'cost': None,
 'datetime': None,
 'fee': None,
 'fees': [],
 'filled': None,
 'id': None,
 'info': {'client_id': '0',
          'parent_order_id': '507088114975509507',
          'parent_order_id_str': '507088114975509507'},
 'lastTradeTimestamp': None,
 'lastUpdateTimestamp': None,
 'postOnly': None,
 'price': None,
 'reduceOnly': None,
 'remaining': None,
 'side': None,
 'status': None,
 'stopLossPrice': None,
 'stopPrice': None,
 'symbol': 'TON/USDT',
 'takeProfitPrice': None,
 'timeInForce': None,
 'timestamp': None,
 'trades': [],
 'triggerPrice': None,
 'type': None}
```

**Review by carlosmiei — APPROVED** [review]

**Pull request merged**